### PR TITLE
[labs/react] Do not override attribtues with event listeners

### LINKED ISSUE #3042: [labs/react] incorrect types without event map
# Description

When a user creates a React component using `createComponent`, attributes are confused for events.

# Steps to Reproduce

1. Write this code

```ts
// simple-greeting.ts

import {html, css, LitElement} from 'lit';
import {customElement, property} from 'lit/decorators.js';
import {createComponent} from '@lit-labs/react';

@customElement('simple-greeting')
export class SimpleGreeting extends LitElement {
  @property()
  name = 'Somebody';

  render() {
    return html`<p>Hello, ${this.name}!</p>`;
  }
}

export const SimpleGreetingReact = createComponent(window.React, 'simple-greeting', SimpleGreeting);
```

and
```
import '@types/react';
import '@types/react-dom';
import {SimpleGreetingReact} from './simple-greeting.js';

const ReactDOM = window.ReactDOM;
const React = window.React;

ReactDOM.render(React.createElement(SimpleGreetingReact, {name: "React"}), document.body);
```

2. See this output...

## Typescript build Error

```
No overload matches this call. The last overload gave the following error. Argument type '{name: string; }' is not assignable to type 'Partial<EventProps<Events> & ..

Property 'name' is incompatible with index signature. Type 'string' is not assignable to type '(e: Event) => void';
```

3. Fix with ...

To suppress error, add `{}` as the event map. Most likely `undefined` is causing some trouble.
```ts
// simple-greeting.ts
...
export const SimpleGreetingReact = createComponent(window.React, 'simple-greeting', SimpleGreeting, {});
```

### PR DESCRIPTION
**Pull request by taylor-vann**
Related to https://github.com/lit/lit/issues/3042

This PR resolves #3042.

The solution was mainly leaving a base case for the generics for `createComponent<element, events>`.
```
...
export const createComponent = <I extends HTMLElement, E extends Events = {}>(
...
```

Previously, leaving out an event map would cause all properties on a component to be considered a generic `event` listener.

With this fix though, only events mapped through `createComponent` args will have unclear types.
Screen Shot 2022-06-21 at 11 11 45 AM

Screen Shot 2022-06-21 at 11 11 54 AM

### CODE REVIEW

**Comment by taylor-vann** [comment]
For future notes:

Typescript inferences two generic types in `createComponent`: `I` and `E`. The `I` will always be defined. It is the third argument of `createComponent` and is required. Typescript will always be able to inference `I` as demonstrated below.

```ts
export const createComponent = <I extends HTMLElement, E extends Events>(
  React: typeof ReactModule,
  tagName: string,
  elementClass: Constructor<I>,
  events?: E,
  ...
```

However the `events` argument or `E` is optional. So typescript will not always be able to inference what `E` is.

Later we define the react component props using `EventProps<E>` which may not have an inference. When we union `&` `React.HTMLAttritbues<HTMLElement> & EventProps<E>` we potentially union an unresolved inference.

```ts
  type UserProps = React.PropsWithChildren<
    React.PropsWithRef<
      Partial<Omit<I, 'children'>> &
        Partial<EventProps<E>> &
        Omit<React.HTMLAttributes<HTMLElement>, keyof E>
    >
  >;
```

This is where i'm a little unclear what happens, but the apparent behavior is when `E` is undefined it will "canabalize" properties from `React.HTMLAttritbues<HTMLElement>`. So the component `props` will be affected by `EventProps` causing property types to default to an event listener `(e: Event) => void`.

Providing an empty object helps resolve the missing inference, when `E` is yet to be inferred it will default to an empty object with no properties. Typescript understands the difference between `React.HTMLAttritbues<HTMLElement>` and `EventProps<E>`.

```ts
export const createComponent = <I extends HTMLElement, E extends Events = {}>(
  ...
```

**Review by sorvell — APPROVED** [review]
As a follow-on, it'd be great if we could add tests that cover this.

**Comment by taylor-vann** [comment]
@sorvell totally agree!

I think the case to test would be more in the Type realm. The bug occurrs when no `Events` map is provided. Rendering a component without events should be enough no? I can add that to this PR as well.

File: .changeset/healthy-cherries-enjoy.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/react': patch
+---
+
+Event map type leak resolved. Component will build without type errors when no event map is provided.
```
Comment (justinfagnani) [inline .changeset/healthy-cherries-enjoy.md:5]: I'd phrase this in a way that makes sense to someone reading a changelog without all the context:

"Fixed an error with typings when no event map is provided to `createComponent()`. Instance properties were being confused with event handlers.

**Review by justinfagnani — CHANGES_REQUESTED** [review]
I really would like to see a test here.

All we need is a component in [here](https://github.com/lit/lit/blob/main/packages/labs/react/src/test/create-component_test.tsx): with a createComponent call without an event map, and a JSX expression that binds a reactive property. We don't really even need any runtime assertions. If it compiles, it's fine.

File: .changeset/healthy-cherries-enjoy.md (hunk 1 above)
Comment (taylor-vann) [reply .changeset/healthy-cherries-enjoy.md:5]: sounds good

Comment (taylor-vann) [reply .changeset/healthy-cherries-enjoy.md]: done

**Comment by taylor-vann** [comment]
The following playground should explain how a `ternaryMap` can cannibalize other object properties in a union.

This package will need to avoid the two breaking cases in this example
https://www.typescriptlang.org/play?#code/PQKgsAUABFDKCuAHApgJygYQPZYDZQBUBPFAZwGNUBLRAFygFUA7KrJqAcS1vIAsBDAISRIMAr2SkqpAFyMWbKIv5QAtv0QoAJlFpom-VEV0lkUcv3ap4VdsgBuaIrV62A5sIghgIiKCiiUABiVKik9ORsUlpoUCpStPD8tKzsAO4SqGYqAErIkahaADzh1ExuADRQpe4AfIHSUPCkyDq0WFC2AGaxKrZ6qF385GZpVC5Y8PTN7lBuyExoVOSkAHS+MBjN7arEZFAAvFAA3oEwAOTk8KhGAPoA1shE53KX13f2-LjwyOdxpJh3kQ9sgSihyFQvrdaKZahUzlBztxMg8ni9Eci0LdPt9fvDoFAAL6QbyQGEoTBRWjWcjtVBFAi1Q4nRZpKAACgAlHICISANxk0yAm7A0wMqDIAAeeiYWgBNXKzIVbiZR2VUAAZCdArdoaYeQKIPzfJBgMBqrRDPQxi44lAAEb8FrmXBO0iQciu0gAgBCTta2oJlt4zIARFokGhQ4aYI6WjojqH7ah+Lxo5BiRBTebmtkoHkCsVlVVlUz7fl+LnhUYQZ0AXo+CwLLhcMZ4tT3IKKQB5KbRNAAo4FrCFEod8ol8duKAAH2ropQtUN2agAE0AJIAaQAorBBIJCK56-pDMYuvAmLTUlAxi2oDFcMg3Mk8zEhvBcPRyWZlExjBfUk8b8oG3aUAAVUCwRBSCKHIJWlBY5SgXtaH7MJVUDGAAG1N06dhHiILAunzABdOQchwkj4JlJCthFEEEQAfnzSisNDXVv1DEiETkJh4FUctUENY0sz8cACQAGWQWhzgBZNkH4e4LXgLoun3TxAikmSAVUSQARtENvwoag6HWAlxEaW98HkxS8N0NIOkQV0RndUSYHZABBTkbyoO8bKU9kMG8+0pigJhuA5AARTkzPcn1vKsh0sls9loodULwvoQKYo2DkvLiWUOXiny-OSpT7WRDkgoKnRUpyrwfFEs08s5FdmEiVRdKYL8JCgLo8FwLAxkVXBbGyQqOq6+hJh6sxFmlXQOn8qAcyYfzsxXSImHCcxkr0P142Ya8jiKddqMQ31-S0Kpu3O2UARQtDSFqdkPSpXaFP2q6jsUE6zqlGjLvjG67qQx6qBiMJmWOQkXsCONWjkbBtupeBaRHU64UCBwFlQxi5G7fFvIOJlTiDIVu1UcY9C0A6AyOSnxlOqoCKI5Cl0CYC6a0H72COMCrUhXAikZ2gae5zUQPAyDoJF2oObcqAskSVB2HZRAZdkKBud54mmS20g8GQVZBrcdXNc5YTfGa9l4ray8sE63HdCPKBRsWGrzEdqalFClw5oBxakoUpTVvWvxGuag36AkFssAJvsIYHaG3pRqBY8GlOCV4IatCIORQwyZIATcHAtFDf55xBfFM2t81sr6i9yA+l8dAMmalDoVIvl0E8jCaBQmFTnaEZ5wegib5lKE+5AdcH22rstyBR95ifL3ZY5R4L-gulDQkl6a+u0vPS8W5pnzbQGAx+4Athh-oFfx6bgAmKe9tn775+5qoM6wA-H9SGvcgz8N5bygOGfgRBQw-1zvncBiBLAWEeKQPeB8gA

**Review by sorvell — DISMISSED** [review]

File: .changeset/healthy-cherries-enjoy.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/react': patch
+---
+
+Fixed an error that occurs when createComponent() is not provided an event map causing instance properties to be confused with event handlers.
```
[the remark this answers is not in the document]
Comment (taylor-vann) [reply .changeset/healthy-cherries-enjoy.md:5]: makes sense! compiling TS, not runtime. Will do!

File: packages/labs/react/src/test/create-component_test.tsx [hunk 3]
```diff
@@ -93,6 +93,24 @@ suite('createComponent', () => {
     await el.updateComplete;
   };
 
+  test('works without event map', async () => {
```
[the remark this answers is not in the document]
Comment (taylor-vann) [reply packages/labs/react/src/test/create-component_test.tsx]: added a comment and gave the test a much better name lol

[the remark this answers is not in the document]
Comment (taylor-vann) [reply packages/labs/react/src/test/create-component_test.tsx]: done!

**Review by justinfagnani — APPROVED** [review]

**Pull request merged**